### PR TITLE
R4R: Refactor mint module

### PR DESCRIPTION
**Pull request by zhiqiang-bianjie**
Resolves: #2089

### CODE REVIEW

File: modules/mint/abci.go [hunk 1]
```diff
@@ -1,22 +1,21 @@
...
-func BeginBlocker(ctx sdk.Context, k Keeper) sdk.Tags {
+func BeginBlocker(ctx sdk.Context, k Keeper) {
 	ctx = ctx.WithLogger(ctx.Logger().With("handler", "beginBlock").With("module", "iris/mint"))
 	logger := ctx.Logger()
 	// Get block BFT time and block height
 	blockTime := ctx.BlockHeader().Time
-	blockHeight := ctx.BlockHeader().Height
 	minter := k.GetMinter(ctx)
-	if blockHeight <= 1 { // don't inflate token in the first block
+	if ctx.BlockHeader().Height <= 1 { // don't inflate token in the first block
```
Comment (chengwenxi) [inline modules/mint/abci.go]: ctx.BlockHeight()

File: config/config.go [hunk 2]
```diff
@@ -1,8 +1,9 @@
 package config
 
 const (
-	Testnet = "testnet"
-	Mainnet = "mainnet"
+	Testnet    = "testnet"
+	Mainnet    = "mainnet"
+	StakeDenom = "iris-atto"
```
Comment (chengwenxi) [inline config/config.go]: `staking.BondDenom` can be modified by `gov`, should consider more.

File: modules/mint/internal/types/minter.go [hunk 3]
```diff
@@ -56,7 +51,8 @@ func (m Minter) NextAnnualProvisions(params Params) (provisions sdk.Dec) {
 }
 
 // get the provisions for a block based on the annual provisions rate
-func (m Minter) BlockProvision(annualProvisions sdk.Dec) sdk.Coin {
-	blockInflationAmount := annualProvisions.QuoInt(sdk.NewInt(blocksPerYear))
-	return sdk.NewCoin(m.MintDenom, blockInflationAmount.TruncateInt())
+func (m Minter) BlockProvision(params Params) sdk.Coin {
```
Comment (chengwenxi) [inline modules/mint/internal/types/minter.go]: I guess it returns the `Next Block Provision`.

File: modules/mint/internal/types/events.go [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package types
+
+const (
+	EventTypeMint = ModuleName
+
+	AttributeKeyLastInflationTime = "last-inflation-time"
+	AttributeKeyInflationTime     = "inflation-time"
+	AttributeKeyMintCoin          = "mint-coin"
```
Comment (chengwenxi) [inline modules/mint/internal/types/events.go]: Use `_` instead of` -`, consistent with SDK.

File: modules/mint/internal/keeper/keeper.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
+package keeper
+
+import (
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/params"
+	"github.com/irisnet/irishub/modules/mint/internal/types"
+)
+
+// keeper of the stake store
```
Comment (chengwenxi) [inline modules/mint/internal/keeper/keeper.go]: stake -> mint

File: modules/mint/internal/types/params.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/params"
+)
+
+var _ params.ParamSet = (*Params)(nil)
+
+// default paramspace for params keeper
+const (
+	DefaultParamSpace = "mint"
+	MintDenom         = "iris-atto"
```
Comment (chengwenxi) [inline modules/mint/internal/types/params.go]: using `sdk.DefaultBondDenom` seems better.

File: modules/mint/internal/types/params.go [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+		{KeyMintDenom, &p.MintDenom},
+	}
+}
+
+func (p Params) String() string {
+	return fmt.Sprintf(`Mint Params:
+  mint/Inflation:  %s,
+  mint/MintDenom:  %s`,
+		p.Inflation.String(), p.MintDenom)
+}
```
Comment (chengwenxi) [inline modules/mint/internal/types/params.go]: unused code

File: simapp/app.go [hunk 8]
```diff
@@ -1,6 +1,7 @@
 package simapp
 
 import (
+	"github.com/irisnet/irishub/modules/mint"
```
Comment (chengwenxi) [inline simapp/app.go]: go import

File: modules/mint/internal/keeper/keeper.go [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+	storeKey         sdk.StoreKey
+	cdc              *codec.Codec
+	paramSpace       params.Subspace
+	supplyKeeper     types.SupplyKeeper
+	feeCollectorName string
+}
+
+func NewKeeper(cdc *codec.Codec, key sdk.StoreKey,
+	paramSpace params.Subspace, supplyKeeper types.SupplyKeeper, feeCollectorName string) Keeper {
+
```
Comment (SegueII) [inline modules/mint/internal/keeper/keeper.go]: ensure module account is set

**Pull request merged**